File: frontend/src/Models/Subunternehmer.ts
import { Fachgebiet } from "../types/Fachgebiet";
import { Rechtsform } from "../types/Rechtsform";

interface SubunternehmerFormData {
  // Informations Générales
  id?: number; // Optionnel côté client si c'est un nouveau formulaire
  firmenname: string; // obligatoire
  ansprechpartnername: string; // obligatoire
  adresse: string; // obligatoire
  telefonnummer: string; // obligatoire
  email: string; // obligatoire

  // Expertise et Qualification
  fachgebiet: Fachgebiet; // obligatoire (choix dans la liste)
  qualifikationen: string; // obligatoire (chemin d'accès au document / description)
  erfarung: string; // obligatoire (courte description)

  // Disponibilité et Région
  verfügbarkeitdatum: string; // obligatoire (Date au format string ISO 8601, ex: "YYYY-MM-DD")
  anzahlDerMitarbeiter: number; // obligatoire
  // Stocké comme une chaîne de caractères séparée par des virgules dans le backend C#, mais géré
  // comme un tableau de chaînes dans le frontend pour la sélection multiple.
  arbeitsregionen: string;

  // Finances et Légal
  umsatznsteuerIdentifikationsnummer: string; // obligatoire
  rechtsformDesUnternehmens: Rechtsform; // obligatoire (choix dans la liste)

  // Assurance (Contient le nom de l'assureur, la couverture et la durée)
  versicherungsschutz: string; // obligatoire

  // Informations Bancaires (Contient le nom et l'IBAN)
  bankverbindung: string; // obligatoire

  // Remarques optionnelles
  sonstige?: string; // optionnel

  // Horodatage de création (généré côté serveur, mais potentiellement lu par le client)
  erstellungsZeit?: string;
}

export default SubunternehmerFormData;

File: backend/Models/Subunternehmer.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema; // Ajouté pour l'attribut Column
using System;
using System.Text.Json.Serialization;
using backend.Services;

namespace backend.Models;



/// <summary>
/// Énumération pour le domaine d'expertise (Fachgebiet) d'un sous-traitant.
/// </summary>
public enum Fachgebiet
{
    Elektrik,
    Rohbau,
    Sanitär,
    Maler,
    Dachdecker
}

/// <summary>
/// Énumération pour la forme juridique de l'entreprise.
/// </summary>
public enum Rechtsform
{
    Einzelunternehmen,
    Personengesellschaften,
    Kapitalgesellschaften
}

/// <summary>
/// Modèle de données pour les sous-traitants (Subunternehmer).
/// </summary>
public class Subunternehmer
{
    [Key]
    public int Id { get; set; }  // Clé primaire

    // Informations Générales
    [Required(ErrorMessage = "Firmenname ist erforderlich")]
    [StringLength(150)]
    public string Firmenname { get; set; } = string.Empty; // obligatoire

    [Required(ErrorMessage = "Ansprechpartnername ist erforderlich")]
    [StringLength(100)]
    public string Ansprechpartnername { get; set; } = string.Empty; // obligatoire

    [Required(ErrorMessage = "Adresse ist erforderlich")]
    public string Adresse { get; set; } = string.Empty; // obligatoire

    [Required(ErrorMessage = "Telefonnummer ist erforderlich")]
    [Phone(ErrorMessage = "Ungültige Telefonnummer")]
    public string Telefonnummer { get; set; } = string.Empty; // obligatoire

    [Required(ErrorMessage = "Email ist erforderlich")]
    [EmailAddress(ErrorMessage = "Ungültige Email-Adresse")]
    public string Email { get; set; } = string.Empty; // obligatoire

    // Expertise et Qualification
    [Required(ErrorMessage = "Fachgebiet ist erforderlich")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public Fachgebiet Fachgebiet { get; set; } // obligatoire (choix dans la liste)

    [Required(ErrorMessage = "Qualifikationen (Dokument/Zertifikat) sind erforderlich")]
    // NOTE: En réalité, ce serait un chemin d'accès à un fichier, mais nous le traitons comme un champ de texte ici.
    public string Qualificationen { get; set; } = string.Empty; // obligatoire (chemin d'accès ou description)

    [Required(ErrorMessage = "Erfahrung ist erforderlich")]
    [StringLength(1000, ErrorMessage = "Beschreibung der Erfahrung darf nicht länger als 1000 Zeichen sein")]
    public string Erfarung { get; set; } = string.Empty; // obligatoire (courte description)

    // Disponibilité et Région
    [Required(ErrorMessage = "Verfügbarkeitsdatum ist erforderlich")]
    [DataType(DataType.Date, ErrorMessage = "Ungültiges Datum")]
    public DateOnly Verfügbarkeitdatum { get; set; } // obligatoire

    [Required(ErrorMessage = "Arbeitsregionen sind erforderlich")]
    // NOTE: Stocké comme une chaîne de caractères séparée par des virgules (ex: "Bayern,Niedersachsen")
   
    public string Arbeitsregionen { get; set; } = string.Empty; // obligatoire (sélection multiple)

    [Required(ErrorMessage = "Anzahl der Mitarbeiter ist erforderlich")]
    public int AnzahlDerMitarbeiter { get; set; } // obligatoire

    // Finances et Légal
    [Required(ErrorMessage = "Umsatzsteuer-Identifikationsnummer ist erforderlich")]
    [StringLength(20)]
    public string UmsatznsteuerIdentifikationsnummer { get; set; } = string.Empty; // obligatoire
    
    [Required(ErrorMessage = "Rechtsform ist erforderlich")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public Rechtsform RechtsformDesUnternehmens { get; set; } // obligatoire (choix dans la liste)

    // Informations d'Assurance (Combinées en un seul champ pour la simplicité du modèle)
    [Required(ErrorMessage = "Angaben zum Versicherungsschutz sind erforderlich")]
    public string Versicherungsschutz { get; set; } = string.Empty; // obligatoire (Nom assureur, Couverture, Durée)

    // Informations Bancaires (Combinées en un seul champ pour la simplicité du modèle)
    [Required(ErrorMessage = "Bankverbindung ist erforderlich")]
    public string Bankverbindung { get; set; } = string.Empty; // obligatoire (Nom et IBAN)
    
    // Remarques optionnelles
    public string Sonstige { get; set; } = string.Empty; // optionnel
    
    // Colonne pour l'horodatage de création (facultatif mais utile)
    public DateTime ErstellungsZeit { get; set; } = DateTime.UtcNow;
}